User

crccw opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/bnuacm/bnuoj-web-v4-rails/7)

<!-- Reviewable:end -->

crccw committed:
mistype

crccw committed:
add locale file for en-us

imported from https://raw.githubusercontent.com/svenfuchs/rails-i18n/master/rails/locale/en-US.yml

crccw committed:
add GET /users/:username

1. Add js-routes to generate url in js, avoiding hardcoded links
2. Add highcharts, it can directly use HTMLElement as data source but fails when there're tags inside cells. So either parsing the table manually, or cloning it then stripping of tags. Currently the latter one is chosen for simplicity.

crccw committed:
links to users when comparing EOM

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/views/contest_list_view.js, line 169 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM4rOB1coOvGF1J_6v)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/contest_list_view.js#L169)):</sup>
we should probably change the return value to numbers to save some bytes

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 19 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM5AZPSmLYfB_coUQv)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L19)):</sup>
use this.$

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 20 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM5FOflqbhjb1u_3TC)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L20)):</sup>
var compareInfoElement, use this.$

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 22 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM4ORA60bCJCGWdDda)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L22)):</sup>
Four spaces.
      var url = Routes.user_compare_path(
          BNUOJ.Utils.getCookie('username'), compare_with);

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
This is not nice. We should avoid to pass html to the client.
Please create a new handlebar template and render it client-side.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 38 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7799ZH7msmO-avv8)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L38)):</sup>
var statTable

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 39 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM74bAKzIyDTtIE3BL)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L39)):</sup>
Wait. Why?

---

<sup>**[app/assets/stylesheets/bnuoj.css.scss, line 70 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM9SjBSQPUVXp_CMck)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/stylesheets/bnuoj.css.scss#L70)):</sup>
Doen't seem right. I don't think we need to do any addtional ui tweak here.

---

<sup>**[app/controllers/users_controller.rb, line 66 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7G1lWoIew_81yx2t)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/controllers/users_controller.rb#L66)):</sup>
no, the path looks wrong to me.
You should try to create a /users/compare/u1/u2

---

<sup>**[app/errors/user_error.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7aWNIDJandnUWPYt)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/errors/user_error.rb#L2)):</sup>
have some text?

---

<sup>**[app/helpers/problems_helper.rb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7g0c54reWwQnSnvD)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/problems_helper.rb#L1)):</sup>
Remove, render it client side. And this is too specific to be a helper.

---

<sup>**[app/helpers/statuses_helper.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7x6TWChGLOz88-Oy)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/statuses_helper.rb#L2)):</sup>
what's hsh?

---

<sup>**[app/models/user.rb, line 75 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM8CX2SuMqBl25TuMl)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/models/user.rb#L75)):</sup>
submitted

---

<sup>**[app/models/user.rb, line 87 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM9KDpl3S_yoAOzgPk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/models/user.rb#L87)):</sup>
counts, no need to abbr this

---

<sup>**[app/models/user.rb, line 88 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM97VShFSzlsJzbxrA)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/models/user.rb#L88)):</sup>
reduce(:+) ?

---

<sup>**[app/views/users/compare.html.erb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM9rvEeUyrLTAH7HPF)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/views/users/compare.html.erb#L1)):</sup>
client side

---

<!-- Sent from Reviewable.io -->

crccw commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/views/contest_list_view.js, line 169 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM4rOB1coOvGF1J_6v)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/contest_list_view.js#L169)):</sup>
which requires modifications of backend, v2, v3... make debug harder while effect might be very limited.. I think we might leave this later..

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 19 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM5AZPSmLYfB_coUQv)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L19)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 20 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM5FOflqbhjb1u_3TC)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L20)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 22 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM4ORA60bCJCGWdDda)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L22)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
I've considered rendering it client-side, but give it up for following reasons:
1. Diffing/Unioning arrays are not as convenient in JS as in RoR, it can be tricky..
2. We won't do any parsing against the html
3. It will cause duplicated code since we have to provide a non-AJAX version for the request.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 38 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7799ZH7msmO-avv8)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L38)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 39 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM74bAKzIyDTtIE3BL)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L39)):</sup>
mentioned in the commit log. Another approach is to parse the table manually.

---

<sup>**[app/assets/stylesheets/bnuoj.css.scss, line 70 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM9SjBSQPUVXp_CMck)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/stylesheets/bnuoj.css.scss#L70)):</sup>
Previously table is used here, cause I think the content does not fit into a table, rules here are to simulate the look, but fine these are not compulsory :)

---

<sup>**[app/controllers/users_controller.rb, line 66 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7G1lWoIew_81yx2t)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/controllers/users_controller.rb#L66)):</sup>
verbs looks not right either...how about /users/comparsion/u1/u2?

BTW /users/:id while id is actually username also seems odd...

---

<sup>**[app/errors/user_error.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7aWNIDJandnUWPYt)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/errors/user_error.rb#L2)):</sup>
huh...you mean move error message to here, so a general error handling could be applied to ApplicationController instead of handling errors individually? 

---

<sup>**[app/helpers/problems_helper.rb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7g0c54reWwQnSnvD)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/problems_helper.rb#L1)):</sup>
How general should a helper be? This exactly same piece of code will be repeated 7 times in user info page. It's ok for me to copy and paste many times, but I just wonder..

---

<sup>**[app/helpers/statuses_helper.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7x6TWChGLOz88-Oy)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/statuses_helper.rb#L2)):</sup>
huh..Now I think this is really no need.

---

<sup>**[app/models/user.rb, line 75 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM8CX2SuMqBl25TuMl)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/models/user.rb#L75)):</sup>
Done.

---

<sup>**[app/models/user.rb, line 87 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM9KDpl3S_yoAOzgPk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/models/user.rb#L87)):</sup>
Done.

---

<sup>**[app/models/user.rb, line 88 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM97VShFSzlsJzbxrA)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/models/user.rb#L88)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

You didn't commit

---

<sup>**[app/assets/javascripts/views/contest_list_view.js, line 169 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM4rOB1coOvGF1J_6v)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/contest_list_view.js#L169)):</sup>
That's not true, this change is v4 only. https://github.com/BNUACM/bnuoj-web-v4-rails/blob/develop/app/models/contest.rb#L71
How would that affect any debugging process? We just change that from server side conversion to client side. You can add a helper if you want.

---

<!-- Sent from Reviewable.io -->

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
1. Not true, http://underscorejs.org/#union http://underscorejs.org/#difference
2. I'm not sure if I understand, why will this affect client side rendering?
3. Why we have to support that? We have plenty of endpoints that will only support json (basically, any ajax endpoint)

---

<sup>**[app/controllers/users_controller.rb, line 66 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7G1lWoIew_81yx2t)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/controllers/users_controller.rb#L66)):</sup>
I don't think we should follow the pattern that everything should end in noun.
This is an action, taking two params.

---

<sup>**[app/errors/user_error.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7aWNIDJandnUWPYt)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/errors/user_error.rb#L2)):</sup>
I mean the error text for each error can be different right?

---

<sup>**[app/helpers/problems_helper.rb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7g0c54reWwQnSnvD)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/problems_helper.rb#L1)):</sup>
Helper supposed to used in different controllers, if I understand this right.
In this case, I think it's only used by a certain controller.

---

<!-- Sent from Reviewable.io -->

crccw commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

Yes. I'm waiting whether to render the page client side

---

<sup>**[app/assets/javascripts/views/contest_list_view.js, line 169 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM4rOB1coOvGF1J_6v)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/contest_list_view.js#L169)):</sup>
huh..Fine, it should be done in another PR.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
1. Cool!
2. So HTML is acceptable
3. Really? I see respond_to in almost every ajax endpoint

---

<sup>**[app/errors/user_error.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7aWNIDJandnUWPYt)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/errors/user_error.rb#L2)):</sup>
Yep and they will be handled differently, I do not do general rescue

---

<sup>**[app/helpers/problems_helper.rb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7g0c54reWwQnSnvD)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/problems_helper.rb#L1)):</sup>
Yep but it ensures uniformity where there should be and makes reading/modifying the view easier. The only disadvantage I've seen is that we might end up with loads of helper methods which can looks chaotic, but codes are either here or in separate views, we make here clear, then views might look chaotic..

---

<!-- Sent from Reviewable.io -->

crccw committed:
fix bugs

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/boot.js, line 11 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkQbcKhNAR22YQ7a7Da)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/boot.js#L11)):</sup>
^users\/[a-zA-Z0-9]+$

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
2. No, it's (almost) always a bad idea to return HTML in ajax. It introduces too much redundant content.
3. I don't think it's true. I never asked for HTML in ajax request.

---

<sup>**[app/helpers/problems_helper.rb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7g0c54reWwQnSnvD)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/problems_helper.rb#L1)):</sup>
It's too specific for me, I can't see any other place that's rendering a list of pids. And it's always a bad idea to assume where the html would be rendered. And why would you have a helper method to generate HTML? That's view's job. If you want to reuse the html, just create a child template.

---

<sup>**[app/helpers/statuses_helper.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7x6TWChGLOz88-Oy)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/statuses_helper.rb#L2)):</sup>
then remove it

---

<!-- Sent from Reviewable.io -->

crccw committed:
Render user_compare client side

2 helpers are provided for Handlebars:
link_to text, route, ...
path_to route, ...

crccw committed:
remove debug info...

crccw committed:
update js routes for /users/:username

crccw committed:
forget a pid_link_list...

crccw commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/boot.js, line 11 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkQbcKhNAR22YQ7a7Da)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/boot.js#L11)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
Done.

---

<sup>**[app/controllers/users_controller.rb, line 66 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7G1lWoIew_81yx2t)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/controllers/users_controller.rb#L66)):</sup>
Done.

---

<sup>**[app/helpers/problems_helper.rb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7g0c54reWwQnSnvD)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/problems_helper.rb#L1)):</sup>
Fine then:)

---

<sup>**[app/helpers/statuses_helper.rb, line 2 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7x6TWChGLOz88-Oy)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/helpers/statuses_helper.rb#L2)):</sup>
Removed

---

<sup>**[app/views/users/compare.html.erb, line 1 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM9rvEeUyrLTAH7HPF)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/views/users/compare.html.erb#L1)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/boot.js, line 11 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkQbcKhNAR22YQ7a7Da)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/boot.js#L11)):</sup>
Align ':'

---

<sup>**[app/assets/javascripts/templates.js, line 14 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFXYlI_PvGDUobSXQ)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L14)):</sup>
var options

---

<sup>**[app/assets/javascripts/templates.js, line 15 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFXEFyPW2ZtUfYqh9)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L15)):</sup>
var args

---

<sup>**[app/assets/javascripts/templates.js, line 20 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFdjbyHGlHMqIAE3V)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L20)):</sup>
ditto

---

<sup>**[app/assets/javascripts/templates.js, line 22 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFkUX7LcTv92botXV)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L22)):</sup>
escapeUri

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 3 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG0hBcMW-A_jr-LjI)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L3)):</sup>
4 spaces indent

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 13 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG7No-6bZR4PnqdkL)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L13)):</sup>
ditto

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 23 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG8KfhwNfMDDlZtSf)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L23)):</sup>
ditto

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 34 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG9CcGBjgJIEm9ICE)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L34)):</sup>
ditto

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 44 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG9yYq5vQ5VFLh3pu)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L44)):</sup>
ditto

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 55 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRGAvlLdDtVgOPzPrT)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L55)):</sup>
ditto

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 19 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM5AZPSmLYfB_coUQv)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L19)):</sup>
instead of calling them partials, we'd probably structure them based on controllers. ie. templates/users/compare

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRGaDjoj9dvZXeE99-)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
Hmmm, I think we should have a "renderTemplate" helper somewhere.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 25 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L25)):</sup>
nit: {
  user1: user1,
  user2: user2
}, data

https://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml?showone=Code_formatting#Code_formatting

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 39 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7799ZH7msmO-avv8)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L39)):</sup>
var statTable

---

<sup>**[app/controllers/users_controller.rb, line 76 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRE3mzTJ4iOKgo5x76)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/controllers/users_controller.rb#L76)):</sup>
We can do this client side right? I'd prefer use raw data and do the calculations client-side. But I don't have strong opinion on this.

---

<sup>**[app/views/users/show.html.erb, line 70 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkREmDVpGw0uLl7Xpjc)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/views/users/show.html.erb#L70)):</sup>
no need to use abbr here, it's just confusing.
Generally, in project we prefer to name the variables with meaningful names. We are not doing competitive programming here :)

---

<!-- Sent from Reviewable.io -->

crccw committed:
add var to local variables

crccw committed:
more fixes...

crccw commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/boot.js, line 11 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkQbcKhNAR22YQ7a7Da)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/boot.js#L11)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates.js, line 14 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFXYlI_PvGDUobSXQ)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L14)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates.js, line 15 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFXEFyPW2ZtUfYqh9)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L15)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates.js, line 20 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFdjbyHGlHMqIAE3V)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L20)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates.js, line 22 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFkUX7LcTv92botXV)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L22)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 3 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG0hBcMW-A_jr-LjI)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L3)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 13 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG7No-6bZR4PnqdkL)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L13)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 23 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG8KfhwNfMDDlZtSf)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L23)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 34 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG9CcGBjgJIEm9ICE)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L34)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 44 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRG9yYq5vQ5VFLh3pu)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L44)):</sup>
Done.

---

<sup>**[app/assets/javascripts/templates/partials/user_compare.jst.hbs, line 55 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRGAvlLdDtVgOPzPrT)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates/partials/user_compare.jst.hbs#L55)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 19 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM5AZPSmLYfB_coUQv)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L19)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 24 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRGaDjoj9dvZXeE99-)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/views/user_show_view.js#L24)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 25 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM6Q99zkYgYgPEhgkk)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L25)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 39 [r2]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkM7799ZH7msmO-avv8)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/e17ba512ddc06bef658cc5b5bcc7be94a2b95bf2/app/assets/javascripts/views/user_show_view.js#L39)):</sup>
Done.

---

<sup>**[app/controllers/users_controller.rb, line 76 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRE3mzTJ4iOKgo5x76)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/controllers/users_controller.rb#L76)):</sup>
Of course, I just think it's more like we've done some comparing, and server is faster...We can directly use GET /users/:id.json if we want to do calculation client-side

---

<sup>**[app/views/users/show.html.erb, line 70 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkREmDVpGw0uLl7Xpjc)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/views/users/show.html.erb#L70)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/templates.js, line 22 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFkUX7LcTv92botXV)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L22)):</sup>
encodeURIComponent
http://xkr.us/articles/javascript/encode-compare/

---

<sup>**[app/assets/javascripts/views/base_view.js, line 133 [r6]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkR_OpT16jby9oX2rVW)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/df49de2b177052d6f0a5ed8a43274f6cef6ac474/app/assets/javascripts/views/base_view.js#L133)):</sup>
add doc

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 25 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRGaDjoj9dvZXeE99-)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/views/user_show_view.js#L25)):</sup>
Change the one in base dialog as well

---

<!-- Sent from Reviewable.io -->

crccw committed:
more and more fixes..

crccw commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/templates.js, line 22 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRFkUX7LcTv92botXV)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/templates.js#L22)):</sup>
Routes contains slashes..

---

<sup>**[app/assets/javascripts/views/base_view.js, line 133 [r6]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkR_OpT16jby9oX2rVW)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/df49de2b177052d6f0a5ed8a43274f6cef6ac474/app/assets/javascripts/views/base_view.js#L133)):</sup>
Done.

---

<sup>**[app/assets/javascripts/views/user_show_view.js, line 25 [r4]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRGaDjoj9dvZXeE99-)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/0d68e2f1c0c856605d8fc6ad77dedfb35fcc6515/app/assets/javascripts/views/user_show_view.js#L25)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->

51isoft commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/views/base_view.js, line 134 [r7]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRenwi3ose3NK4LI6h)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/7c48c7c12c69e8f863c5f8ed0978faf438735357/app/assets/javascripts/views/base_view.js#L134)):</sup>
targetElement

---

<!-- Sent from Reviewable.io -->

crccw committed:
targetElement..

crccw committed:
base_dialog.js

crccw commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7)

---

<sup>**[app/assets/javascripts/views/base_view.js, line 134 [r7]](https://reviewable.io:443/reviews/bnuacm/bnuoj-web-v4-rails/7#-JkRenwi3ose3NK4LI6h)** ([raw file](https://github.com/bnuacm/bnuoj-web-v4-rails/blob/7c48c7c12c69e8f863c5f8ed0978faf438735357/app/assets/javascripts/views/base_view.js#L134)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->
